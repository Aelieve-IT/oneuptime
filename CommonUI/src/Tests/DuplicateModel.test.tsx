import React from 'react';
import BaseModel from 'Common/Models/BaseModel';
import TableMetaData from 'Common/Types/Database/TableMetadata';
import IconProp from 'Common/Types/Icon/IconProp';
import { render, screen, within, fireEvent, waitFor } from '@testing-library/react';
import DuplicateModel from '../Components/DuplicateModel/DuplicateModel';
import ObjectID from 'Common/Types/ObjectID';
import CrudApiEndpoint from 'Common/Types/Database/CrudApiEndpoint';
import { act } from 'react-test-renderer';
import Route from 'Common/Types/API/Route';

@TableMetaData({
    tableName: 'Foo',
    singularName: 'Foo',
    pluralName: 'Foos',
    icon: IconProp.Wrench,
    tableDescription: 'A test model'
})
@CrudApiEndpoint(new Route('/testModel'))
class TestModel extends BaseModel {
    public changeThis?: string = 'original';
};

jest.mock('../Utils/ModelAPI/ModelAPI', () => {
    return {
        getItem: jest.fn().mockResolvedValueOnce({
            changeThis: 'changed',
            setValue: function (key:'changeThis', value: string) {
                this[key] = value;
            },
            removeValue: jest.fn(),
        }).mockResolvedValueOnce({
            changeThis: 'changed',
            setValue: function (key:'changeThis', value: string) {
                this[key] = value;
            },
            removeValue: jest.fn(),
        }).mockResolvedValueOnce(undefined),
        create: jest.fn().mockResolvedValueOnce({
            data: {
                id: 'foobar',
                changeThis: 'changed',
            },
        }).mockResolvedValueOnce(undefined),
    };
});

jest.mock('../Utils/Navigation',() => {
    return {
        navigate: jest.fn()
    }
});

describe('DuplicateModel', () => {
    const fieldsToDuplicate = {};
    const fieldsToChange = [{
        field: {
            changeThis: true
        },
        title: 'Change This',
        required: false,
        placeholder: 'You can change this'
    }];
    it('renders correctly', () => {
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
            />
        );
        expect(document.body).toMatchSnapshot();
    });
    it('shows confirmation modal when duplicate button is clicked',() => {
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
            />
        );
        const button = screen.getByRole('button',{name: 'Duplicate Foo'});
        fireEvent.click(button);
        expect(screen.getByRole('dialog')).toMatchSnapshot();
    });
    it('duplicates item when confirmation button is clicked',async () => {
        const onDuplicateSuccess = jest.fn();
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
                onDuplicateSuccess={onDuplicateSuccess}
                navigateToOnSuccess={new Route('/done')}
            />
        );
        const button = screen.getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(button);
        });
        const dialog = screen.getByRole('dialog');
        const confirmationButton = within(dialog).getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(confirmationButton);
        });
        await waitFor(
            () => expect(
                onDuplicateSuccess
            ).toBeCalledWith(
                {
                    id: 'foobar',
                    changeThis: 'changed',
                },
            )
        );
        await waitFor(
            () => expect(
                require('../Utils/Navigation').navigate
            ).toBeCalledWith(
                new Route('/done/foobar')
            )
        );
    });
    it('closes confirmation dialog when close button is clicked',() => {
        const onDuplicateSuccess = jest.fn();
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
                onDuplicateSuccess={onDuplicateSuccess}
                navigateToOnSuccess={new Route('/done')}
            />
        );
        const button = screen.getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(button);
        });
        const dialog = screen.getByRole('dialog');
        const closeButton = within(dialog).getByRole('button',{name: 'Close'});
        act(() => {
            fireEvent.click(closeButton);
        });
        expect(screen.queryByRole('dialog')).toBeFalsy();
    });
    it('handles could not create error correctly',async () => {
        const onDuplicateSuccess = jest.fn();
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
                onDuplicateSuccess={onDuplicateSuccess}
                navigateToOnSuccess={new Route('/done')}
            />
        );
        const button = screen.getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(button);
        });
        const dialog = screen.getByRole('dialog');
        const confirmationButton = within(dialog).getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(confirmationButton);
        });
        await screen.findByText('Duplicate Error');
        expect(screen.getByRole('dialog')).toMatchSnapshot();
    });
    it('handles item not found error correctly',async () => {
        const onDuplicateSuccess = jest.fn();
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
                onDuplicateSuccess={onDuplicateSuccess}
                navigateToOnSuccess={new Route('/done')}
            />
        );
        const button = screen.getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(button);
        });
        const dialog = screen.getByRole('dialog');
        const confirmationButton = within(dialog).getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(confirmationButton);
        });
        await screen.findByText('Duplicate Error');
        expect(screen.getByRole('dialog')).toMatchSnapshot();
    });
    it('closes error dialog when close button is clicked',async () => {
        const onDuplicateSuccess = jest.fn();
        render(
            <DuplicateModel
                modelType={TestModel}
                modelId={new ObjectID('foo')}
                fieldsToDuplicate={fieldsToDuplicate}
                fieldsToChange={fieldsToChange}
                onDuplicateSuccess={onDuplicateSuccess}
                navigateToOnSuccess={new Route('/done')}
            />
        );
        const button = screen.getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(button);
        });
        const dialog = screen.getByRole('dialog');
        const confirmationButton = within(dialog).getByRole('button',{name: 'Duplicate Foo'});
        act(() => {
            fireEvent.click(confirmationButton);
        });
        await screen.findByText('Duplicate Error');
        const errorDialog = screen.getByRole('dialog');
        const closeButton = within(errorDialog).getByRole('button',{name: 'Close'});
        act(() => {
            fireEvent.click(closeButton);
        });
        expect(screen.queryByRole('dialog')).toBeFalsy();
    });
});