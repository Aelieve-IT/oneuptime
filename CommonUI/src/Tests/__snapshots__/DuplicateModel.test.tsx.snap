// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DuplicateModel handles could not create error correctly 1`] = `
<div
  aria-labelledby="modal-title"
  aria-modal="true"
  class="relative z-20"
  role="dialog"
>
  <div
    class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
  />
  <div
    class="fixed inset-0 z-20 overflow-y-auto"
  >
    <div
      class="flex min-h-full items-end justify-center p-4 text-center sm:items-center sm:p-0"
    >
      <div
        class="relative transform rounded-lg bg-white text-left shadow-xl transition-all sm:my-8 sm:w-full   sm:max-w-lg "
      >
        <div
          class="sm:p-6"
        >
          <div
            class="mt-3 text-center sm:mt-0 sm:ml-4 sm:mr-4 sm:text-left"
          >
            <div
              class="flex justify-between"
            >
              <div>
                <h3
                  class="text-lg font-medium leading-6 text-gray-900 "
                  id="modal-title"
                >
                  Duplicate Error
                </h3>
              </div>
            </div>
            <div
              class="mt-2"
            >
              <div
                class="modal-body"
              >
                <div
                  class="text-gray-500 mt-5 text-sm"
                >
                  Error: Could not create Foo
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="bg-gray-50 px-4 py-3 sm:flex sm:flex-row-reverse sm:px-6"
        >
          <button
            class="inline-flex w-full justify-center rounded-md border border-gray-300 bg-white text-base font-medium text-gray-700 shadow-sm hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 sm:mt-0 sm:ml-3 sm:w-auto sm:text-sm px-3 py-2"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DuplicateModel handles item not found error correctly 1`] = `
<div
  aria-labelledby="modal-title"
  aria-modal="true"
  class="relative z-20"
  role="dialog"
>
  <div
    class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
  />
  <div
    class="fixed inset-0 z-20 overflow-y-auto"
  >
    <div
      class="flex min-h-full items-end justify-center p-4 text-center sm:items-center sm:p-0"
    >
      <div
        class="relative transform rounded-lg bg-white text-left shadow-xl transition-all sm:my-8 sm:w-full   sm:max-w-lg "
      >
        <div
          class="sm:p-6"
        >
          <div
            class="mt-3 text-center sm:mt-0 sm:ml-4 sm:mr-4 sm:text-left"
          >
            <div
              class="flex justify-between"
            >
              <div>
                <h3
                  class="text-lg font-medium leading-6 text-gray-900 "
                  id="modal-title"
                >
                  Duplicate Error
                </h3>
              </div>
            </div>
            <div
              class="mt-2"
            >
              <div
                class="modal-body"
              >
                <div
                  class="text-gray-500 mt-5 text-sm"
                >
                  Error: Could not find Foo with id foo
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="bg-gray-50 px-4 py-3 sm:flex sm:flex-row-reverse sm:px-6"
        >
          <button
            class="inline-flex w-full justify-center rounded-md border border-gray-300 bg-white text-base font-medium text-gray-700 shadow-sm hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 sm:mt-0 sm:ml-3 sm:w-auto sm:text-sm px-3 py-2"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DuplicateModel renders correctly 1`] = `
<body>
  <div>
    <section>
      <div
        class="shadow sm:overflow-hidden sm:rounded-md"
      >
        <div
          class="bg-white py-6 px-4 sm:p-6"
        >
          <div
            class="flex justify-between"
          >
            <div>
              <h2
                class="text-lg font-medium leading-6 text-gray-900"
                id="payment-details-heading"
              >
                Duplicate Foo
              </h2>
              <p
                class="mt-1 text-sm text-gray-500"
              >
                Duplicating this foo will create another foo exactly like this one.
              </p>
            </div>
            <div
              class="flex"
            >
              <div>
                <button
                  class="inline-flex w-full justify-center rounded-md border border-gray-300 bg-white text-base font-medium text-gray-700 shadow-sm hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 sm:mt-0 sm:ml-3 sm:w-auto sm:text-sm px-3 py-2"
                  type="button"
                >
                  <div
                    role="icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="   w-5 h-5 mr-1"
                      fill="none"
                      stroke="currentColor"
                      stroke-width="1.5"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M15.75 17.25v3.375c0 .621-.504 1.125-1.125 1.125h-9.75a1.125 1.125 0 01-1.125-1.125V7.875c0-.621.504-1.125 1.125-1.125H6.75a9.06 9.06 0 011.5.124m7.5 10.376h3.375c.621 0 1.125-.504 1.125-1.125V11.25c0-4.46-3.243-8.161-7.5-8.876a9.06 9.06 0 00-1.5-.124H9.375c-.621 0-1.125.504-1.125 1.125v3.5m7.5 10.375H9.375a1.125 1.125 0 01-1.125-1.125v-9.25m12 6.625v-1.875a3.375 3.375 0 00-3.375-3.375h-1.5a1.125 1.125 0 01-1.125-1.125v-1.5a3.375 3.375 0 00-3.375-3.375H9.75"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                      />
                    </svg>
                  </div>
                  Duplicate Foo
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;

exports[`DuplicateModel shows confirmation modal when duplicate button is clicked 1`] = `
<div
  aria-labelledby="modal-title"
  aria-modal="true"
  class="relative z-20"
  role="dialog"
>
  <div
    class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
  />
  <div
    class="fixed inset-0 z-20 overflow-y-auto"
  >
    <div
      class="flex min-h-full items-end justify-center p-4 text-center sm:items-center sm:p-0"
    >
      <div
        class="relative transform rounded-lg bg-white text-left shadow-xl transition-all sm:my-8 sm:w-full   sm:max-w-lg "
      >
        <div
          class="absolute top-0 right-0 hidden pt-4 pr-4 sm:block"
        >
          <button
            class="rounded-md bg-white text-gray-600   focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 px-3 py-2"
            type="button"
          >
            <div
              role="icon"
            >
              <svg
                aria-hidden="true"
                class="   w-5 h-5 m-1"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 18L18 6M6 6l12 12"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          class="sm:p-6"
        >
          <div
            class="mt-3 text-center sm:mt-0 sm:ml-4 sm:mr-4 sm:text-left"
          >
            <div
              class="flex justify-between"
            >
              <div>
                <h3
                  class="text-lg font-medium leading-6 text-gray-900 "
                  id="modal-title"
                >
                  Duplicate Foo
                </h3>
                <h3
                  class="text-sm leading-6 text-gray-500"
                >
                  Are you sure you want to duplicate this foo?
                </h3>
              </div>
            </div>
            <div
              class="mt-2"
            >
              <div
                class="modal-body"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-lg-1"
                  >
                    <div>
                      <div
                        class="flex"
                      >
                        <div
                          class="w-full pt-1"
                        >
                          <div>
                            <div
                              class="grid md:grid-cols-1 grid-cols-1 gap-4"
                            >
                              <div>
                                <div
                                  class="sm:col-span-4 mt-0 mb-2"
                                >
                                  <label
                                    class="block text-sm  font-medium text-gray-700 flex justify-between"
                                  >
                                    <span>
                                      Change This
                                       
                                      <span
                                        class="text-gray-400 text-xs"
                                      >
                                        (Optional)
                                      </span>
                                    </span>
                                  </label>
                                  <div
                                    class="mt-2"
                                  />
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="flex w-full justify-end"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="bg-gray-50 px-4 py-3 sm:flex sm:flex-row-reverse sm:px-6"
        >
          <button
            class="inline-flex w-full justify-center rounded-md border border-transparent bg-indigo-600 text-base font-medium text-white shadow-sm  focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 sm:ml-3 sm:w-auto sm:text-sm px-3 py-2"
            type="submit"
          >
            Duplicate Foo
          </button>
          <button
            class="inline-flex w-full justify-center rounded-md border border-gray-300 bg-white text-base font-medium text-gray-700 shadow-sm hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 sm:mt-0 sm:ml-3 sm:w-auto sm:text-sm px-3 py-2"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
